### ISSUE TITLE
Forms: SKCanvasView in ListView glitch

### ISSUE THREAD

**Issue by LeRondPoint** [mono/SkiaSharp#304]
I got an issue with a SKCanvasView when used in a list.  The list is dynamically resized and after many many many redraws, the contexts seem to get messed up and they swap from one row to the other.  So in the following sample, just tap on the names repeatedly (around 20-30 times should do the trick).  My guess is that the bug is alwas there but the amount of redraws makes it easier to reproduce in the below sample.

https://github.com/LeRondPoint/FormsSkiaBikeTracker/tree/SKCanvasViewBug

Thanks

**Comment by mattleibow** [comment]
This one has been a bit tough to look at. I think I have seen something about what you mention, but I may need more info.

First, what device are you testing on? I tested an iPhone 6s with iOS 10.2

You say the contexts get swapped - what are you meaning here? could you get some screenshot of a bad case?

I did manage to get some weirdness, when tapping various items repeatedly, two things happened:

1. The text would randomly change size
2. The text would flicker and sometimes disappear for a few milliseconds.

These two issue were mainly cause by the fact that the size of the canvas (SKLabel) was changing as the row expanded or collapsed. This issue disappeared when I set a fixed height to the label element.

Other then those two issues, I am not sure exactly what you may mean, more info will help a great deal.

**Comment by LeRondPoint** [comment]
Those two issues are what I'm refering to.  The names also get swapped at times, but always keep the same View width, stretching or compressing the contents.  It looks like the wrong memory buffer is being assigned to the Canvas.
Tested on iphone 5s 9.3 simulator, iPod touch 4th gen ios 9.3 and iPhone 5 ios 10.2.

I have a way to reproduce it almost all the time if it helps.  You need to change the SKLabel to the DrawableView base class and rename a few methods (see the CircledImage as a sample).  I'm not exactly sure what is different between both custom renderers to be honest but they behave differently.  If you want the code for the DrawableView I can provide it, given it helps.

Is there any other way I can help on this?

**Comment by LeRondPoint** [comment]
Using the DrawableView base class, I end up with this 👍 
```
// **********************************************************************
// 
//   SKLabel.xaml.cs
//   
//   This file is subject to the terms and conditions defined in
//   file 'LICENSE.txt', which is part of this source code package.
//   
//   Copyright (c) 2017, Le rond-point
// 
// ***********************************************************************

using System;
using System.IO;
using LRPLib.Services.Resources;
using LRPLib.Views.XForms;
using MvvmCross.Platform;
using SkiaSharp;
using SkiaSharp.Views.Forms;
using Xamarin.Forms;

namespace FormsSkiaBikeTracker.Forms.UI.Controls
{
    public class SKLabel : DrawableView
    {
        public static readonly BindableProperty FontResourcePathProperty = BindableProperty.Create(nameof(FontResourcePath), typeof(string), typeof(SKLabel), string.Empty, BindingMode.OneWay, null, FontResourcePathPropertyChanged);
        public static readonly BindableProperty TextProperty = BindableProperty.Create(nameof(Text), typeof(string), typeof(SKLabel), string.Empty, BindingMode.OneWay, null, ResizePropertyChanged);
        public static readonly BindableProperty TextColorProperty = BindableProperty.Create(nameof(TextColor), typeof(Color), typeof(SKLabel), Color.Black, BindingMode.OneWay, null, InvalidatePropertyChanged);
        public static readonly BindableProperty FontSizeProperty = BindableProperty.Create(nameof(FontSize), typeof(double), typeof(SKLabel), 14.0, BindingMode.OneWay, null, ResizePropertyChanged);
        private IResourceLocator _resourceLocator;

        public string FontResourcePath
        {
            get { return (string)GetValue(FontResourcePathProperty); }
            set { SetValue(FontResourcePathProperty, value); }
        }

        public string Text
        {
            get { return (string)GetValue(TextProperty); }
            set { SetValue(TextProperty, value); }
        }

        public Color TextColor
        {
            get { return (Color)GetValue(TextColorProperty); }
            set { SetValue(TextColorProperty, value); }
        }

        public double FontSize
        {
            get { return (double)GetValue(FontSizeProperty); }
            set { SetValue(FontSizeProperty, value); }
        }

        private SizeRequest? _LastSize { get; set; }
        private SKPaint _Paint { get; }
        private SKTypeface _Typeface { get; set; }

        public SKLabel()
        {
            _Paint = new SKPaint();
            _resourceLocator = Mvx.Resolve<IResourceLocator>();
        }

        ~SKLabel()
        {
            _Typeface?.Dispose();
            _Paint?.Dispose();

            HasTransparency = true;
        }

        private static void FontResourcePathPropertyChanged(BindableObject bindable, object oldvalue, object newvalue)
        {
            SKLabel view = bindable as SKLabel;

            view.RefreshPaint();
            view.RefreshTypeface();
            ResizePropertyChanged(bindable, oldvalue, newvalue);
        }

        private void RefreshTypeface()
        {
            string resourcePath = _resourceLocator.GetResourcePath("Fonts", FontResourcePath);
            Stream resStream = _resourceLocator.ResourcesAssembly.GetManifestResourceStream(resourcePath);

            _Typeface?.Dispose();

            if (resStream != null)
            {
                using (resStream)
                {
                    _Typeface = SKTypeface.FromStream(resStream);
                }
            }
            else
            {
                _Typeface = null;
            }
        }

        private static void InvalidatePropertyChanged(BindableObject bindable, object oldValue, object newValue)
        {
            SKLabel view = bindable as SKLabel;

            view.RefreshPaint();
            view.Invalidate();
        }
                
        private static void ResizePropertyChanged(BindableObject bindable, object oldValue, object newValue)
        {
            SKLabel view = bindable as SKLabel;

            view.RefreshPaint();
            view.InvalidateMeasure();
            view.Invalidate();
        }

        protected override void InvalidateMeasure()
        {
            _LastSize = null;
            base.InvalidateMeasure();
        }

        protected override SizeRequest OnMeasure(double widthConstraint, double heightConstraint)
        {
            if (_LastSize == null && HasSomethingToDraw())
            {
                SKRect textBounds = new SKRect();
                Rectangle requestRect;

                _Paint.MeasureText(Text, ref textBounds);
                requestRect = textBounds.ToFormsRect();

                requestRect.Height += _Paint.FontMetrics.Descent;
                requestRect.Width = Math.Ceiling(Math.Min(requestRect.Width, widthConstraint));
                requestRect.Height = Math.Ceiling(Math.Min(requestRect.Height, heightConstraint));

                _LastSize = new SizeRequest(requestRect.Size);
            }

            return _LastSize.GetValueOrDefault(new SizeRequest());
        }

        protected override void Paint(SKCanvas canvas)
        {
            if (HasSomethingToDraw())
            {
                canvas.DrawText(Text, 0, (float)Height - (_Paint.FontMetrics.Descent * 0.5f), _Paint);
            }
        }

        private bool HasSomethingToDraw()
        {
            return _Typeface != null && !string.IsNullOrEmpty(Text) && TextColor != Color.Transparent && FontSize > 0;
        }

        private void RefreshPaint()
        {
            _Paint.Typeface = _Typeface;
            _Paint.Color = TextColor.ToSKColor();
            _Paint.TextSize = (float)FontSize;
            _Paint.IsAntialias = true;
        }
    }
}
```

And bug...

![image](https://cloud.githubusercontent.com/assets/15251145/26493661/ca0cedb4-41e8-11e7-9d7e-cd55d730f739.png)

**Comment by LeRondPoint** [comment]
Hello,
I updated my sample repo to make it easier to isolate the issue.  I removed my custom renderer and am now using the SKCanvasView with a bit of helpers and getting the glitch big time.  My CrawableView class is very simple : 

```
using SkiaSharp;
using SkiaSharp.Views.Forms;

namespace LRPLib.Views.XForms
{
    public abstract class DrawableView : SKCanvasView
    {
        public bool HasTransparency { get; set; }
        public SKPath ClippingPath { get; set; }

        protected override void OnPaintSurface(SKPaintSurfaceEventArgs e)
        {
            SKCanvas canvas = e.Surface.Canvas;
            float screenDensity = e.Info.Width / (float)Width;

            canvas.Scale(screenDensity, screenDensity);
             
            if (HasTransparency || ClippingPath != null)
            {
                canvas.Clear();
            }

            if (ClippingPath != null)
            {
                canvas.ClipPath(ClippingPath);
            }

            Paint(canvas);
        }

        protected void Invalidate()
        {
            InvalidateSurface();
        }

        protected abstract void Paint(SKCanvas canvas);
    }
}
```

**Comment by LeRondPoint** [comment]
Alright so after more investigation, I realized my issue was due to having the Skia bitmaps and Canvas in the renderer instead of in a view.  Since Xamarin reuses renderers in a list, the draw data was being passed from one row to the other, hence causing the glitches.  After using the Skia provided views and renderers and removing possibilities of small resizes due to floating point errors in my OnMeasure, all seems well.

Closing this issue.
